flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/MapUnionFunction.java

```diff
@@ -0,0 +1,100 @@
+ */
+
+package org.apache.flink.table.runtime.functions.scalar;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.data.ArrayData;
+import org.apache.flink.table.data.GenericArrayData;
+import org.apache.flink.table.data.MapData;
+import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
+import org.apache.flink.table.functions.SpecializedFunction;
+import org.apache.flink.table.types.KeyValueDataType;
+
+import javax.annotation.Nullable;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/** Implementation of {@link BuiltInFunctionDefinitions#MAP_UNION}. */
+@Internal
+public class MapUnionFunction extends BuiltInScalarFunction {
+
+    private final ArrayData.ElementGetter keyElementGetter;
+    private final ArrayData.ElementGetter valueElementGetter;
+
+    public MapUnionFunction(SpecializedFunction.SpecializedContext context) {
+        super(BuiltInFunctionDefinitions.MAP_UNION, context);
+        KeyValueDataType outputType =
+                ((KeyValueDataType) context.getCallContext().getOutputDataType().get());
+        keyElementGetter =
+                ArrayData.createElementGetter(outputType.getKeyDataType().getLogicalType());
+        valueElementGetter =
+                ArrayData.createElementGetter(outputType.getValueDataType().getLogicalType());
+    }
+
+    public @Nullable MapData eval(@Nullable MapData map1, @Nullable MapData map2) {
+        if (map1 == null || map2 == null) {
+            return null;
+        }
+        MapDataForMapUnion resultMap = new MapDataForMapUnion(map1, map2);
+        return resultMap;
+    }
+
+    private class MapDataForMapUnion implements MapData {
+        private ArrayData keysArray;
+        private ArrayData valuesArray;
+        private final Map<Object, Object> map;
+
+        public MapDataForMapUnion(MapData map1, MapData map2) {
+
+            this.map = new HashMap<>();
+            for (int i = 0; i < map1.size(); i++) {
+                final Object key = keyElementGetter.getElementOrNull(map1.keyArray(), i);
+                final Object value = valueElementGetter.getElementOrNull(map1.valueArray(), i);
+                this.map.put(key, value);
+            }
+            for (int i = 0; i < map2.size(); i++) {
+                final Object key = keyElementGetter.getElementOrNull(map2.keyArray(), i);
+                final Object value = valueElementGetter.getElementOrNull(map2.valueArray(), i);
+                this.map.put(key, value);
+            }
+            this.updateArrays();
+        }
+
+        @Override
+        public int size() {
+            return map.size();
+        }
+
+        @Override
+        public ArrayData keyArray() {
+            return keysArray;
+        }
+
+        @Override
+        public ArrayData valueArray() {
+            return valuesArray;
+        }
+
+        private void updateArrays() {
+            this.keysArray = new GenericArrayData(map.keySet().toArray());
```

Comment:
do we need a separate method for that if it's used only once and contains only 2 lines?

Reply:
If we use equalityEvaluator, we will update map many times. So I think we should keep it now.

Reply:
fixed it.